File: Medical_Profile/Enc256.cs
using JWT;
using JWT.Algorithms;
using JWT.Exceptions;
using JWT.Serializers;
using System;
using System.Collections.Generic;
using System.IO;
using System.Security.Cryptography;
using System.Text;

namespace Enc
{
  static class Enc256
  {
    private readonly static RandomNumberGenerator Random = RandomNumberGenerator.Create();

    public readonly static int BlockBitSize = 128;
    public readonly static int KeyBitSize = 256;
    public readonly static int SaltBitSize = 64;
    public readonly static int Iterations = 10000;
    public readonly static int MinPasswordLength = 12;
  public static string Iterscramble(string inp, int iter = 1)
  {
   string pstring = inp;
   for (int i = 0; i <= iter; i++)
    pstring = Scramble(pstring);
   return pstring;
  } 

  public static int Next_open(byte[] tg, int st)
    {
      int p = st + 1;

      if (p >= tg.Length)
        p = 0;

      while (true)
      {
        if (tg[p] == 0xFF)
        {
          break;
        }
        p += 1;
        if (p == tg.Length)
        {
          p = 0;
        }
      }
      return p;
    }

    public static string Scramble(string inp)
    {
      byte[] b1 = new byte[inp.Length];
      byte[] b2 = new byte[inp.Length];

      for (int i = 0; i < b1.Length; i++)
      {
        b1[i] = (byte)i;
      }

      for (int j = 0; j < b2.Length; j++)
      {
        b2[j] = 0xFF;
      }

      for(int i = 0;i<b2.Length;i++)
      {
        b2[Next_open(b2, i+i)] = (byte)inp[i]; 
      }
      return System.Text.Encoding.Default.GetString(b2);
    }

    public static string Encode(string secret, string salt, Dictionary<string, object> payload, int iter)
    {
      IJwtAlgorithm algorithm = new HMACSHA256Algorithm();
      IJsonSerializer serializer = new JsonNetSerializer();
      IBase64UrlEncoder urlEncoder = new JwtBase64UrlEncoder();
      IJwtEncoder encoder = new JwtEncoder(algorithm, serializer, urlEncoder);

      var token = encoder.Encode(payload, NewKey(secret, salt, iter));

      return (token);
    }

    public static string Encode(string secret, string Salt, Dictionary<string, object> payload)
    {
      IJwtAlgorithm algorithm = new HMACSHA256Algorithm();
      IJsonSerializer serializer = new JsonNetSerializer();
      IBase64UrlEncoder urlEncoder = new JwtBase64UrlEncoder();
      IJwtEncoder encoder = new JwtEncoder(algorithm, serializer, urlEncoder);

      byte[] saltb = Convert.FromBase64String(Salt);

      int iter = (int)Math.Abs(Convert.ToInt64((BitConverter.ToUInt64(saltb, 0) + 10000) % 10000));

      var token = encoder.Encode(payload, NewKey(secret, Salt, iter));

      token = Salt + "." + token;

      return (token);
    }

    public static IDictionary<string, object> Decode(string token, string secret, string salt, int iter)
    {
      IDictionary<string, object> rdict = new Dictionary<string, object>();
      try
      {
        IJsonSerializer serializer = new JsonNetSerializer();
        IDateTimeProvider provider = new UtcDateTimeProvider();
        IJwtValidator validator = new JwtValidator(serializer, provider);
        IBase64UrlEncoder urlEncoder = new JwtBase64UrlEncoder();
        IJwtDecoder decoder = new JwtDecoder(serializer, validator, urlEncoder,new HMACSHA256Algorithm());

        var dict = decoder.DecodeToObject<IDictionary<string, object>>(token, NewKey(secret, salt, iter), verify: true);

        return (dict);
      }
      catch (TokenExpiredException)
      {
        rdict["Error"] = "Token has expired";
        return rdict;
      }
      catch (SignatureVerificationException)
      {
        rdict["Error"] = "Token has invalid signature";
        return rdict;
      }
    }

    public static IDictionary<string, object> Decode(string token, string secret)
    {
      var i = token.IndexOf(".");

      string salt = token.Substring(0, i);

      string Stoken = token.Substring(i + 1);

      byte[] saltb = Convert.FromBase64String(salt);

      int iter = (int)Math.Abs(Convert.ToInt64((BitConverter.ToUInt64(saltb, 0) + 10000) % 10000));

      return (Decode(Stoken, secret, salt, iter));


    }

    public static string NewKey(string password, string salt, int Iterations = 10000)
    {
      var saltb = Convert.FromBase64String(salt);
      byte[] key = new byte[KeyBitSize / 8];
      using (var generator = new Rfc2898DeriveBytes(password, saltb, Iterations))
      {
        key = generator.GetBytes(KeyBitSize / 8);
      }

      return Convert.ToBase64String(key, 0, key.Length);
    }

    public static string SimpleEncrypt(string secretMessage, byte[] cryptKey, byte[] authKey, byte[] nonSecretPayload = null)
    {
      if (string.IsNullOrEmpty(secretMessage))
      {
        throw new ArgumentException("Secret Message Required!", "secretMessage");
      }

      var plainText = Encoding.UTF8.GetBytes(secretMessage);
      var cipherText = SimpleEncrypt(plainText, cryptKey, authKey, nonSecretPayload);
      return Convert.ToBase64String(cipherText);
    }

    public static string SimpleDecrypt(string encryptedMessage, byte[] cryptKey, byte[] authKey, int nonSecretPayloadLength = 0)
    {
      if (string.IsNullOrWhiteSpace(encryptedMessage))
      {
        throw new ArgumentException("Encrypted Message Required!", "encryptedMessage");
      }

      var cipherText = Convert.FromBase64String(encryptedMessage);
      var plainText = SimpleDecrypt(cipherText, cryptKey, authKey, nonSecretPayloadLength);
      return plainText == null ? null : Encoding.UTF8.GetString(plainText);
    }

    public static string Encrypt(string secretmessage, string password, string Label)
    {
      int v1 = 0;
      int l1 = Convert.ToInt32(Label.Substring(3));

      for (int i = 3; i < Label.Length; i++)
      {
        v1 += Convert.ToInt32(Label.Substring(i, 1));
      }

      int insert_loc = v1 % (Label.Length - 4);

      string k1 = password.Substring(0, insert_loc) + Label + password.Substring(insert_loc);

      return (Encrypt(secretmessage, k1, l1));
    }
    public static string Encrypt(string secretMessage, string password, int iter = 10000, byte[] nonSecretPayload = null)
    {
      if (string.IsNullOrEmpty(secretMessage))
      {
        throw new ArgumentException("Secret Message Required!", "secretMessage");
      }

      var plainText = Encoding.UTF8.GetBytes(secretMessage);
      var cipherText = SimpleEncryptWithPassword(plainText, password, iter, nonSecretPayload);
      return Convert.ToBase64String(cipherText);
    }


    public static string SimpleEncryptWithPassword(string secretMessage, string password, int iter, byte[] nonSecretPayload = null)
    {
      if (string.IsNullOrEmpty(secretMessage))
      {
        throw new ArgumentException("Secret Message Required!", "secretMessage");
      }

      var plainText = Encoding.UTF8.GetBytes(secretMessage);
      var cipherText = SimpleEncryptWithPassword(plainText, password, iter, nonSecretPayload);
      return Convert.ToBase64String(cipherText);
    }

    public static string Decrypt(string secretmessage, string password, string Label)
    {
      int v1 = 0;
      int l1 = Convert.ToInt32(Label.Substring(3));

      for (int i = 3; i < Label.Length; i++)
      {
        v1 += Convert.ToInt32(Label.Substring(i, 1));
      }

      int insert_loc = v1 % (Label.Length - 4);

      string k1 = password.Substring(0, insert_loc) + Label + password.Substring(insert_loc);

      return (Decrypt(secretmessage, k1, l1));
    }
    public static string Decrypt(string encryptedMessage, string password, int iter = 10000, int nonSecretPayloadLength = 0)
    {
      if (string.IsNullOrWhiteSpace(encryptedMessage))
      {
        throw new ArgumentException("Encrypted Message Required!", "encryptedMessage");
      }

      var cipherText = Convert.FromBase64String(encryptedMessage);
      var plainText = SimpleDecryptWithPassword(cipherText, password, iter, nonSecretPayloadLength);
      return plainText == null ? null : Encoding.UTF8.GetString(plainText);
    }

    public static string SimpleDecryptWithPassword(string encryptedMessage, string password, int iter = 10000, int nonSecretPayloadLength = 0)
    {
      if (string.IsNullOrWhiteSpace(encryptedMessage))
      {
        throw new ArgumentException("Encrypted Message Required!", "encryptedMessage");
      }

      var cipherText = Convert.FromBase64String(encryptedMessage);
      var plainText = SimpleDecryptWithPassword(cipherText, password, nonSecretPayloadLength);
      return plainText == null ? null : Encoding.UTF8.GetString(plainText);
    }

    public static byte[] SimpleEncrypt(byte[] secretMessage, byte[] cryptKey, byte[] authKey, byte[] nonSecretPayload = null)
    {
      if (cryptKey == null || cryptKey.Length != KeyBitSize / (double)8)
      {
        throw new ArgumentException(string.Format("Key needs to be {0} bit!", KeyBitSize), "cryptKey");
      }

      if (authKey == null || authKey.Length != KeyBitSize / (double)8)
      {
        throw new ArgumentException(string.Format("Key needs to be {0} bit!", KeyBitSize), "authKey");
      }

      if (secretMessage == null || secretMessage.Length < 1)
      {
        throw new ArgumentException("Secret Message Required!", "secretMessage");
      }

      nonSecretPayload = nonSecretPayload ?? new byte[] { };
      byte[] cipherText;
      byte[] iv;

      using (var aes = new AesManaged() { KeySize = KeyBitSize, BlockSize = BlockBitSize, Mode = CipherMode.CBC, Padding = PaddingMode.PKCS7 })
      {
        aes.GenerateIV();
        iv = aes.IV;

        using (var encrypter = aes.CreateEncryptor(cryptKey, iv))
        {
          using (var cipherStream = new MemoryStream())
          {
            using (var cryptoStream = new CryptoStream(cipherStream, encrypter, CryptoStreamMode.Write))
            {
              using (var binaryWriter = new BinaryWriter(cryptoStream))
              {
                binaryWriter.Write(secretMessage);
              }
            }

            cipherText = cipherStream.ToArray();
          }
        }
      }

      using (var hmac = new HMACSHA256(authKey))
      {
        using (var encryptedStream = new MemoryStream())
        {
          using (var binaryWriter = new BinaryWriter(encryptedStream))
          {
            binaryWriter.Write(nonSecretPayload);
            binaryWriter.Write(iv);
            binaryWriter.Write(cipherText);
            binaryWriter.Flush();
            var tag = hmac.ComputeHash(encryptedStream.ToArray());
            binaryWriter.Write(tag);
          }

          return encryptedStream.ToArray();
        }
      }
    }

    public static byte[] SimpleDecrypt(byte[] encryptedMessage, byte[] cryptKey, byte[] authKey, int nonSecretPayloadLength = 0)
    {
      if (cryptKey == null || cryptKey.Length != KeyBitSize / (double)8)
      {
        throw new ArgumentException(string.Format("CryptKey needs to be {0} bit!", KeyBitSize), "cryptKey");
      }

      if (authKey == null || authKey.Length != KeyBitSize / (double)8)
      {
        throw new ArgumentException(string.Format("AuthKey needs to be {0} bit!", KeyBitSize), "authKey");
      }

      if (encryptedMessage == null || encryptedMessage.Length == 0)
      {
        throw new ArgumentException("Encrypted Message Required!", "encryptedMessage");
      }

      using (var hmac = new HMACSHA256(authKey))
      {
        var sentTag = new byte[hmac.HashSize / 8 - 1 + 1];
        var calcTag = hmac.ComputeHash(encryptedMessage, 0, encryptedMessage.Length - sentTag.Length);
        var ivLength = (BlockBitSize / 8);
        if (encryptedMessage.Length < sentTag.Length + nonSecretPayloadLength + ivLength)
        {
          return null;
        }

        Array.Copy(encryptedMessage, encryptedMessage.Length - sentTag.Length, sentTag, 0, sentTag.Length);
        var compare = 0;
        var loopTo = sentTag.Length - 1;
        for (var i = 0; i <= loopTo; i++)
        {
          compare = compare | sentTag[i] ^ calcTag[i];
        }

        if (compare != 0)
        {
          return null;
        }

        using (var aes = new AesManaged() { KeySize = KeyBitSize, BlockSize = BlockBitSize, Mode = CipherMode.CBC, Padding = PaddingMode.PKCS7 })
        {
          var iv = new byte[ivLength - 1 + 1];
          Array.Copy(encryptedMessage, nonSecretPayloadLength, iv, 0, iv.Length);

          using (var decrypter = aes.CreateDecryptor(cryptKey, iv))
          {
            using (var plainTextStream = new MemoryStream())
            {
              using (var decrypterStream = new CryptoStream(plainTextStream, decrypter, CryptoStreamMode.Write))
              {
                using (var binaryWriter = new BinaryWriter(decrypterStream))
                {
                  binaryWriter.Write(encryptedMessage, nonSecretPayloadLength + iv.Length, encryptedMessage.Length - nonSecretPayloadLength - iv.Length - sentTag.Length);
                }
              }

              return plainTextStream.ToArray();
            }
          }
        }
      }
    }

    public static byte[] SimpleEncryptWithPassword(byte[] secretMessage, string password, int iter = 10000, byte[] nonSecretPayload = null)
    {
      nonSecretPayload = nonSecretPayload ?? new byte[] { };
      if (string.IsNullOrWhiteSpace(password) || password.Length < MinPasswordLength)
      {
        throw new ArgumentException(string.Format("Must have a password of at least {0} characters!", MinPasswordLength), "password");
      }

      if (secretMessage == null || secretMessage.Length == 0)
      {
        throw new ArgumentException("Secret Message Required!", "secretMessage");
      }

      var payload = new byte[((SaltBitSize / 8) * 2) + nonSecretPayload.Length - 1 + 1];
      Array.Copy(nonSecretPayload, payload, nonSecretPayload.Length);
      int payloadIndex = nonSecretPayload.Length;
      byte[] cryptKey;
      byte[] authKey;

      using (var generator = new Rfc2898DeriveBytes(password, SaltBitSize / 8, Iterations))
      {
        var salt = generator.Salt;
        cryptKey = generator.GetBytes(KeyBitSize / 8);
        Array.Copy(salt, 0, payload, payloadIndex, salt.Length);
        payloadIndex += salt.Length;
      }

      using (var generator = new Rfc2898DeriveBytes(password, SaltBitSize / 8, Iterations))
      {
        var salt = generator.Salt;
        authKey = generator.GetBytes(KeyBitSize / 8);
        Array.Copy(salt, 0, payload, payloadIndex, salt.Length);
      }

      return SimpleEncrypt(secretMessage, cryptKey, authKey, payload);
    }

    public static string Getsalt(string message)
    {
      if (string.IsNullOrWhiteSpace(message))
      {
        throw new ArgumentException("Encrypted Message Required!", "message");
      }

      var cText = Convert.FromBase64String(message);
      var Salt = new byte[SaltBitSize / 8 - 1 + 1];
      Array.Copy(cText, 0, Salt, 0, Salt.Length);
      return Convert.ToBase64String(Salt, 0, Salt.Length);
    }

    public static byte[] SimpleDecryptWithPassword(byte[] encryptedMessage, string password, int iter = 10000, int nonSecretPayloadLength = 0)
    {
      if (string.IsNullOrWhiteSpace(password) || password.Length < MinPasswordLength)
      {
        throw new ArgumentException(string.Format("Must have a password of at least {0} characters!", MinPasswordLength), "password");
      }

      if (encryptedMessage == null || encryptedMessage.Length == 0)
      {
        throw new ArgumentException("Encrypted Message Required!", "encryptedMessage");
      }

      var cryptSalt = new byte[SaltBitSize / 8 - 1 + 1];
      var authSalt = new byte[SaltBitSize / 8 - 1 + 1];
      Array.Copy(encryptedMessage, nonSecretPayloadLength, cryptSalt, 0, cryptSalt.Length);
      Array.Copy(encryptedMessage, nonSecretPayloadLength + cryptSalt.Length, authSalt, 0, authSalt.Length);
      byte[] cryptKey;
      byte[] authKey;

      using (var generator = new Rfc2898DeriveBytes(password, cryptSalt, Iterations))
      {
        cryptKey = generator.GetBytes(KeyBitSize / 8);
      }

      using (var generator = new Rfc2898DeriveBytes(password, authSalt, Iterations))
      {
        authKey = generator.GetBytes(KeyBitSize / 8);
      }

      return SimpleDecrypt(encryptedMessage, cryptKey, authKey, cryptSalt.Length + authSalt.Length + nonSecretPayloadLength);
    }
  }
}

File: Medical_Profile/Enc256.vb
Imports JWT
Imports JWT.Algorithms
Imports JWT.Serializers
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Security.Cryptography
Imports System.Text

Namespace Enc
 Friend Module Enc256
  Private ReadOnly Random As RandomNumberGenerator = RandomNumberGenerator.Create()
  Public ReadOnly BlockBitSize As Integer = 128
  Public ReadOnly KeyBitSize As Integer = 256
  Public ReadOnly SaltBitSize As Integer = 64
  Public ReadOnly Iterations As Integer = 10000
  Public ReadOnly MinPasswordLength As Integer = 12

  Function Encode(ByVal secret As String, ByVal salt As String, ByVal payload As Dictionary(Of String, Object), ByVal iter As Integer) As String
   Dim algorithm As IJwtAlgorithm = New HMACSHA256Algorithm()
   Dim serializer As IJsonSerializer = New JsonNetSerializer()
   Dim urlEncoder As IBase64UrlEncoder = New JwtBase64UrlEncoder()
   Dim encoder As IJwtEncoder = New JwtEncoder(algorithm, serializer, urlEncoder)
   Dim token = encoder.Encode(payload, NewKey(secret, salt, iter))
   Return (token)
  End Function

  Function Encode(ByVal secret As String, ByVal Salt As String, ByVal payload As Dictionary(Of String, Object)) As String
   Dim algorithm As IJwtAlgorithm = New HMACSHA256Algorithm()
   Dim serializer As IJsonSerializer = New JsonNetSerializer()
   Dim urlEncoder As IBase64UrlEncoder = New JwtBase64UrlEncoder()
   Dim encoder As IJwtEncoder = New JwtEncoder(algorithm, serializer, urlEncoder)
   Dim saltb As Byte() = Convert.FromBase64String(Salt)
   Dim iter As Integer = CInt(Math.Abs(Convert.ToInt64((BitConverter.ToUInt64(saltb, 0) + 10000) Mod 10000)))
   Dim token = encoder.Encode(payload, NewKey(secret, Salt, iter))
   token = Salt & "." & token
   Return (token)
  End Function

  Function Decode(ByVal token As String, ByVal secret As String, ByVal salt As String, ByVal iter As Integer) As IDictionary(Of String, Object)
   Dim rdict As IDictionary(Of String, Object) = New Dictionary(Of String, Object)()

   Try
    Dim serializer As IJsonSerializer = New JsonNetSerializer()
    Dim provider As IDateTimeProvider = New UtcDateTimeProvider()
    Dim validator As IJwtValidator = New JwtValidator(serializer, provider)
    Dim urlEncoder As IBase64UrlEncoder = New JwtBase64UrlEncoder()
    Dim decoder As IJwtDecoder = New JwtDecoder(serializer, validator, urlEncoder)
    Dim dict = decoder.DecodeToObject(Of IDictionary(Of String, Object))(token, NewKey(secret, salt, iter), verify:=True)
    Return (dict)
   Catch __unusedTokenExpiredException1__ As TokenExpiredException
    rdict("Error") = "Token has expired"
    Return rdict
   Catch __unusedSignatureVerificationException2__ As SignatureVerificationException
    rdict("Error") = "Token has invalid signature"
    Return rdict
   End Try
  End Function

  Function Decode(ByVal token As String, ByVal secret As String) As IDictionary(Of String, Object)
   Dim i = token.IndexOf(".")
   Dim salt As String = token.Substring(0, i)
   Dim Stoken As String = token.Substring(i + 1)
   Dim saltb As Byte() = Convert.FromBase64String(salt)
   Dim iter As Integer = CInt(Math.Abs(Convert.ToInt64((BitConverter.ToUInt64(saltb, 0) + 10000) Mod 10000)))
   Return (Decode(Stoken, secret, salt, iter))
  End Function

    Function NewKey(ByVal password As String, ByVal salt As String, ByVal Optional Iterations As Integer = 10000) As String
      Dim saltb = Convert.FromBase64String(salt)
      Dim key As Byte() = New Byte(KeyBitSize / 8 - 1) {}

      Using generator = New Rfc2898DeriveBytes(password, saltb, Iterations)
        key = generator.GetBytes(KeyBitSize / 8)
      End Using

      Return Convert.ToBase64String(key, 0, key.Length)
    End Function

    Function Next_open(ByVal tg As Byte(), ByVal st As Integer) As Integer
      Dim p As Integer = st + 1
      If p >= tg.Length Then p = 0

      While True

        If tg(p) = &HFF Then
          Exit While
        End If

        p += 1

        If p = tg.Length Then
          p = 0
        End If
      End While

      Return p
    End Function

    Function Iterscramble(ByVal inp As String, Optional ByVal iter As Integer = 1) As String
      Dim pstring As String = inp
      For i As Integer = 0 To iter
        pstring = Scramble(pstring)
      Next
      Return pstring
    End Function

    Function Scramble(ByVal inp As String) As String
      Dim b1 As Byte() = New Byte(inp.Length - 1) {}
      Dim b2 As Byte() = New Byte(inp.Length - 1) {}
      Dim car() As Char = inp.ToCharArray

      For i As Integer = 0 To b1.Length - 1
        b1(i) = CByte(i)
      Next

      For j As Integer = 0 To b2.Length - 1
        b2(j) = &HFF
      Next

      For i As Integer = 0 To b2.Length - 1
        b2(Next_open(b2, i + i)) = Convert.ToByte(car(i))
      Next

      Return System.Text.Encoding.[Default].GetString(b2)
    End Function

    Function SimpleEncrypt(ByVal secretMessage As String, ByVal cryptKey As Byte(), ByVal authKey As Byte(), ByVal Optional nonSecretPayload As Byte() = Nothing) As String
   If String.IsNullOrEmpty(secretMessage) Then
    Throw New ArgumentException("Secret Message Required!", "secretMessage")
   End If

   Dim plainText = Encoding.UTF8.GetBytes(secretMessage)
   Dim cipherText = SimpleEncrypt(plainText, cryptKey, authKey, nonSecretPayload)
   Return Convert.ToBase64String(cipherText)
  End Function

  Function SimpleDecrypt(ByVal encryptedMessage As String, ByVal cryptKey As Byte(), ByVal authKey As Byte(), ByVal Optional nonSecretPayloadLength As Integer = 0) As String
   If String.IsNullOrWhiteSpace(encryptedMessage) Then
    Throw New ArgumentException("Encrypted Message Required!", "encryptedMessage")
   End If

   Dim cipherText = Convert.FromBase64String(encryptedMessage)
   Dim plainText = SimpleDecrypt(cipherText, cryptKey, authKey, nonSecretPayloadLength)
   Return If(plainText Is Nothing, Nothing, Encoding.UTF8.GetString(plainText))
  End Function

  Function Encrypt(ByVal secretmessage As String, ByVal password As String, ByVal Label As String) As String
   Dim v1 As Integer = 0
   Dim l1 As Integer = Convert.ToInt32(Label.Substring(3))

   For i As Integer = 3 To Label.Length - 1
    v1 += Convert.ToInt32(Label.Substring(i, 1))
   Next

   Dim insert_loc As Integer = v1 Mod (Label.Length - 4)
   Dim k1 As String = password.Substring(0, insert_loc) & Label & password.Substring(insert_loc)
   Return (Encrypt(secretmessage, k1, l1))
  End Function

  Function Encrypt(ByVal secretMessage As String, ByVal password As String, ByVal Optional iter As Integer = 10000, ByVal Optional nonSecretPayload As Byte() = Nothing) As String
   If String.IsNullOrEmpty(secretMessage) Then
    Throw New ArgumentException("Secret Message Required!", "secretMessage")
   End If

   Dim plainText = Encoding.UTF8.GetBytes(secretMessage)
   Dim cipherText = SimpleEncryptWithPassword(plainText, password, iter, nonSecretPayload)
   Return Convert.ToBase64String(cipherText)
  End Function

  Function SimpleEncryptWithPassword(ByVal secretMessage As String, ByVal password As String, ByVal iter As Integer, ByVal Optional nonSecretPayload As Byte() = Nothing) As String
   If String.IsNullOrEmpty(secretMessage) Then
    Throw New ArgumentException("Secret Message Required!", "secretMessage")
   End If

   Dim plainText = Encoding.UTF8.GetBytes(secretMessage)
   Dim cipherText = SimpleEncryptWithPassword(plainText, password, iter, nonSecretPayload)
   Return Convert.ToBase64String(cipherText)
  End Function

  Function Decrypt(ByVal secretmessage As String, ByVal password As String, ByVal Label As String) As String
   Dim v1 As Integer = 0
   Dim l1 As Integer = Convert.ToInt32(Label.Substring(3))

   For i As Integer = 3 To Label.Length - 1
    v1 += Convert.ToInt32(Label.Substring(i, 1))
   Next

   Dim insert_loc As Integer = v1 Mod (Label.Length - 4)
   Dim k1 As String = password.Substring(0, insert_loc) & Label & password.Substring(insert_loc)
   Return (Decrypt(secretmessage, k1, l1))
  End Function

  Function Decrypt(ByVal encryptedMessage As String, ByVal password As String, ByVal Optional iter As Integer = 10000, ByVal Optional nonSecretPayloadLength As Integer = 0) As String
   If String.IsNullOrWhiteSpace(encryptedMessage) Then
    Throw New ArgumentException("Encrypted Message Required!", "encryptedMessage")
   End If

   Dim cipherText = Convert.FromBase64String(encryptedMessage)
   Dim plainText = SimpleDecryptWithPassword(cipherText, password, iter, nonSecretPayloadLength)
   Return If(plainText Is Nothing, Nothing, Encoding.UTF8.GetString(plainText))
  End Function

  Function SimpleDecryptWithPassword(ByVal encryptedMessage As String, ByVal password As String, ByVal Optional iter As Integer = 10000, ByVal Optional nonSecretPayloadLength As Integer = 0) As String
   If String.IsNullOrWhiteSpace(encryptedMessage) Then
    Throw New ArgumentException("Encrypted Message Required!", "encryptedMessage")
   End If

   Dim cipherText = Convert.FromBase64String(encryptedMessage)
   Dim plainText = SimpleDecryptWithPassword(cipherText, password, nonSecretPayloadLength)
   Return If(plainText Is Nothing, Nothing, Encoding.UTF8.GetString(plainText))
  End Function

  Function SimpleEncrypt(ByVal secretMessage As Byte(), ByVal cryptKey As Byte(), ByVal authKey As Byte(), ByVal Optional nonSecretPayload As Byte() = Nothing) As Byte()
   If cryptKey Is Nothing OrElse cryptKey.Length <> KeyBitSize / CDbl(8) Then
    Throw New ArgumentException(String.Format("Key needs to be {0} bit!", KeyBitSize), "cryptKey")
   End If

   If authKey Is Nothing OrElse authKey.Length <> KeyBitSize / CDbl(8) Then
    Throw New ArgumentException(String.Format("Key needs to be {0} bit!", KeyBitSize), "authKey")
   End If

   If secretMessage Is Nothing OrElse secretMessage.Length < 1 Then
    Throw New ArgumentException("Secret Message Required!", "secretMessage")
   End If

   nonSecretPayload = If(nonSecretPayload, New Byte() {})
   Dim cipherText As Byte()
   Dim iv As Byte()

   Using aes = New AesManaged() With {
       .KeySize = KeyBitSize,
       .BlockSize = BlockBitSize,
       .Mode = CipherMode.CBC,
       .Padding = PaddingMode.PKCS7
   }
    aes.GenerateIV()
    iv = aes.IV

    Using encrypter = aes.CreateEncryptor(cryptKey, iv)

     Using cipherStream = New MemoryStream()

      Using cryptoStream = New CryptoStream(cipherStream, encrypter, CryptoStreamMode.Write)

       Using binaryWriter = New BinaryWriter(cryptoStream)
        binaryWriter.Write(secretMessage)
       End Using
      End Using

      cipherText = cipherStream.ToArray()
     End Using
    End Using
   End Using

   Using hmac = New HMACSHA256(authKey)

    Using encryptedStream = New MemoryStream()

     Using binaryWriter = New BinaryWriter(encryptedStream)
      binaryWriter.Write(nonSecretPayload)
      binaryWriter.Write(iv)
      binaryWriter.Write(cipherText)
      binaryWriter.Flush()
      Dim tag = hmac.ComputeHash(encryptedStream.ToArray())
      binaryWriter.Write(tag)
     End Using

     Return encryptedStream.ToArray()
    End Using
   End Using
  End Function

  Function SimpleDecrypt(ByVal encryptedMessage As Byte(), ByVal cryptKey As Byte(), ByVal authKey As Byte(), ByVal Optional nonSecretPayloadLength As Integer = 0) As Byte()
   If cryptKey Is Nothing OrElse cryptKey.Length <> KeyBitSize / CDbl(8) Then
    Throw New ArgumentException(String.Format("CryptKey needs to be {0} bit!", KeyBitSize), "cryptKey")
   End If

   If authKey Is Nothing OrElse authKey.Length <> KeyBitSize / CDbl(8) Then
    Throw New ArgumentException(String.Format("AuthKey needs to be {0} bit!", KeyBitSize), "authKey")
   End If

   If encryptedMessage Is Nothing OrElse encryptedMessage.Length = 0 Then
    Throw New ArgumentException("Encrypted Message Required!", "encryptedMessage")
   End If

   Using hmac = New HMACSHA256(authKey)
    Dim sentTag = New Byte(hmac.HashSize / 8 - 1 + 1 - 1) {}
    Dim calcTag = hmac.ComputeHash(encryptedMessage, 0, encryptedMessage.Length - sentTag.Length)
    Dim ivLength = (BlockBitSize / 8)

    If encryptedMessage.Length < sentTag.Length + nonSecretPayloadLength + ivLength Then
     Return Nothing
    End If

    Array.Copy(encryptedMessage, encryptedMessage.Length - sentTag.Length, sentTag, 0, sentTag.Length)
    Dim compare = 0
    Dim loopTo = sentTag.Length - 1

    For i = 0 To loopTo
     compare = compare Or sentTag(i) Xor calcTag(i)
    Next

    If compare <> 0 Then
     Return Nothing
    End If

    Using aes = New AesManaged() With {
        .KeySize = KeyBitSize,
        .BlockSize = BlockBitSize,
        .Mode = CipherMode.CBC,
        .Padding = PaddingMode.PKCS7
    }
     Dim iv = New Byte(ivLength - 1 + 1 - 1) {}
     Array.Copy(encryptedMessage, nonSecretPayloadLength, iv, 0, iv.Length)

     Using decrypter = aes.CreateDecryptor(cryptKey, iv)

      Using plainTextStream = New MemoryStream()

       Using decrypterStream = New CryptoStream(plainTextStream, decrypter, CryptoStreamMode.Write)

        Using binaryWriter = New BinaryWriter(decrypterStream)
         binaryWriter.Write(encryptedMessage, nonSecretPayloadLength + iv.Length, encryptedMessage.Length - nonSecretPayloadLength - iv.Length - sentTag.Length)
        End Using
       End Using

       Return plainTextStream.ToArray()
      End Using
     End Using
    End Using
   End Using
  End Function

  Function SimpleEncryptWithPassword(ByVal secretMessage As Byte(), ByVal password As String, ByVal Optional iter As Integer = 10000, ByVal Optional nonSecretPayload As Byte() = Nothing) As Byte()
   nonSecretPayload = If(nonSecretPayload, New Byte() {})

   If String.IsNullOrWhiteSpace(password) OrElse password.Length < MinPasswordLength Then
    Throw New ArgumentException(String.Format("Must have a password of at least {0} characters!", MinPasswordLength), "password")
   End If

   If secretMessage Is Nothing OrElse secretMessage.Length = 0 Then
    Throw New ArgumentException("Secret Message Required!", "secretMessage")
   End If

   Dim payload = New Byte(((SaltBitSize / 8) * 2) + nonSecretPayload.Length - 1 + 1 - 1) {}
   Array.Copy(nonSecretPayload, payload, nonSecretPayload.Length)
   Dim payloadIndex As Integer = nonSecretPayload.Length
   Dim cryptKey As Byte()
   Dim authKey As Byte()

   Using generator = New Rfc2898DeriveBytes(password, SaltBitSize / 8, Iterations)
    Dim salt = generator.Salt
    cryptKey = generator.GetBytes(KeyBitSize / 8)
    Array.Copy(salt, 0, payload, payloadIndex, salt.Length)
    payloadIndex += salt.Length
   End Using

   Using generator = New Rfc2898DeriveBytes(password, SaltBitSize / 8, Iterations)
    Dim salt = generator.Salt
    authKey = generator.GetBytes(KeyBitSize / 8)
    Array.Copy(salt, 0, payload, payloadIndex, salt.Length)
   End Using

   Return SimpleEncrypt(secretMessage, cryptKey, authKey, payload)
  End Function

  Function Getsalt(ByVal message As String) As String
   If String.IsNullOrWhiteSpace(message) Then
    Throw New ArgumentException("Encrypted Message Required!", "message")
   End If

   Dim cText = Convert.FromBase64String(message)
   Dim Salt = New Byte(SaltBitSize / 8 - 1 + 1 - 1) {}
   Array.Copy(cText, 0, Salt, 0, Salt.Length)
   Return Convert.ToBase64String(Salt, 0, Salt.Length)
  End Function

  Function SimpleDecryptWithPassword(ByVal encryptedMessage As Byte(), ByVal password As String, ByVal Optional iter As Integer = 10000, ByVal Optional nonSecretPayloadLength As Integer = 0) As Byte()
   If String.IsNullOrWhiteSpace(password) OrElse password.Length < MinPasswordLength Then
    Throw New ArgumentException(String.Format("Must have a password of at least {0} characters!", MinPasswordLength), "password")
   End If

   If encryptedMessage Is Nothing OrElse encryptedMessage.Length = 0 Then
    Throw New ArgumentException("Encrypted Message Required!", "encryptedMessage")
   End If

   Dim cryptSalt = New Byte(SaltBitSize / 8 - 1 + 1 - 1) {}
   Dim authSalt = New Byte(SaltBitSize / 8 - 1 + 1 - 1) {}
   Array.Copy(encryptedMessage, nonSecretPayloadLength, cryptSalt, 0, cryptSalt.Length)
   Array.Copy(encryptedMessage, nonSecretPayloadLength + cryptSalt.Length, authSalt, 0, authSalt.Length)
   Dim cryptKey As Byte()
   Dim authKey As Byte()

   Using generator = New Rfc2898DeriveBytes(password, cryptSalt, Iterations)
    cryptKey = generator.GetBytes(KeyBitSize / 8)
   End Using

   Using generator = New Rfc2898DeriveBytes(password, authSalt, Iterations)
    authKey = generator.GetBytes(KeyBitSize / 8)
   End Using

   Return SimpleDecrypt(encryptedMessage, cryptKey, authKey, cryptSalt.Length + authSalt.Length + nonSecretPayloadLength)
  End Function
 End Module
End Namespace

